src/main/java/org/dbos/apiary/mysql/MysqlContext.java

```diff
@@ -86,9 +88,29 @@ public void executeUpsert(String tableName, String id, Object... input) throws E
         // make writes visible
         String updateVisibility = String.format("UPDATE %s SET %s = ? WHERE %s = ? AND %s < ? AND %s = ?", tableName, MysqlContext.endVersion, MysqlContext.apiaryID, MysqlContext.beginVersion, MysqlContext.endVersion);
         // UPDATE table SET __endVersion__ = ? WHERE __apiaryID__ = ? and __beginVersion__ < ? and __endVersion__ == infinity
-        pstmt = conn.prepareStatement(updateVisibility);
-        prepareStatement(pstmt, new Object[]{txc.txID, id, txc.txID, Long.MAX_VALUE});
-        pstmt.executeUpdate();
+        if (visbilityUpdatePredicate != null) {
+            updateVisibility = updateVisibility + " AND " + visbilityUpdatePredicate;
+        }
+        while (true) {
```

Comment:
What does this do?  I assume this is to make sure the visibility update never fails?  Is it possible you get an unrecoverable error here and loop infinitely?